Lib/traceback.py

```
@@ -1504,24 +1504,27 @@ def _compute_suggestion_error(exc_value, tb, wrong_name):
             return None
     else:
         assert isinstance(exc_value, NameError)
-        # find most recent frame
-        if tb is None:
+        try:
```

This is a very broad try-except block. If you scoped it only around the hasattr call would the new test cases still pass?

---

Yes they would still pass.  It's usually good practice to minimise code in a `try` block, but I think this is an exceptional case.  This function attempts to provide "Did you mean foo?" suggestions after there has already been an error in user code.  Hence a different exception propagating from this function is very unhelpful (and the cause of the issue).  So I think it's worth having wider `try` blocks here to minimise the risk of unhandled exceptions in the future.  The existing code has wide `try` blocks ignoring all Exception subclasses when doing processing for AttributeError (L1485-1494) and ImportError (L1499-1502).  I think we we should treat the processing for NameError the same.  (Although - I haven't found a way to cause an exception in L1508-1517, so this is mostly theoretical.)  Happy to narrow the scope of the `try` if folks feel strongly.